File: CPP/Solutions/2015/4/Year2015Day04.cpp
#include <utility>
#include <sstream>
#include <iostream>
#include <algorithm>
#include <numeric>
#include <set>
#include <tuple>

#include "Year2015Day04.h"
#include "md5.h"

auto Year2015Day04::parse() {
    return this->input[0];
}

std::string Year2015Day04::part1() {
    std::string input = this->parse();
    MD5 md5;
    int idx = 100000;
    std::string sb = md5(input + std::to_string(idx)).substr(0, 5);
    while(sb != "00000") {
        idx++;
        sb = md5(input + std::to_string(idx)).substr(0, 5);
    }
    return std::to_string(idx);
}

std::string Year2015Day04::part2() {
    std::string input = this->parse();
    MD5 md5;
    int idx = 1000000;
    std::string sb = md5(input + std::to_string(idx)).substr(0, 6);
    while(sb != "000000") {
        idx++;
        sb = md5(input + std::to_string(idx)).substr(0, 6);
    }
    return std::to_string(idx);
}

Year2015Day04::Year2015Day04(std::vector<std::string> lines) {
    this->input = std::move(lines);
}

File: CSharp/Solutions/2015/4/Year2015Day04.cs
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using AdventOfCode.Common;

namespace AdventOfCode.Solutions._2015._4
{
	// ReSharper disable once UnusedMember.Global
	public class Year2015Day04 : ISolution
	{
		public string Part1(IEnumerable<string> input)
		{
			var key = input.First();
			var number = 1;
			while (true)
			{
				var hash = CreateMd5($"{key}{number}");
				if(hash.StartsWith("00000"))
					break;
				number++;
			}
			return number.ToString();
		}

		public string Part2(IEnumerable<string> input)
		{
			var key = input.First();
			var number = 1;
			while (true)
			{
				var hash = CreateMd5($"{key}{number}");
				if(hash.StartsWith("000000"))
					break;
				number++;
			}
			return number.ToString();
		}

		private static string CreateMd5(string input)
		{
			using (var md5 = MD5.Create())
			{
				var inputBytes = Encoding.ASCII.GetBytes(input);
				var hashBytes = md5.ComputeHash(inputBytes);
				var sb = new StringBuilder();
				foreach (var hashByte in hashBytes)
				{
					sb.Append(hashByte.ToString("X2"));
				}

				return sb.ToString();
			}
		}

	}
}